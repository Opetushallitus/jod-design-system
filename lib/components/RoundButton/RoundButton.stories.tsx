import type { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';
import { RoundButton } from './RoundButton';

const meta = {
  title: 'Primitives/RoundButton',
  component: RoundButton,
  tags: ['autodocs'],
} satisfies Meta<typeof RoundButton>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Base: Story = {
  args: {
    label: 'Goals',
    onClick: fn(),
    icon: 'target',
  },
};
