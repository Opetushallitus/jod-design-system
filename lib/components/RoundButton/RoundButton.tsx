import { tidyClasses } from '../../utils';

export interface RoundButtonProps {
  /** Text shown on the button */
  label: string;
  /** Hide label */
  hideLabel?: boolean;
  /** Callback fired on tap/click of the button */
  onClick: () => void;
  /** Button disabled for any actions */
  disabled?: boolean;
  /** Selected */
  selected?: boolean;
  /** CSS classname */
  className?: string;
  /** Icon shown on the link */
  icon: React.ReactNode;
  /** Icon size: sm = 32px, md = 40px, lg = 64px  */
  size?: 'sm' | 'md' | 'lg';
  /** Background color */
  bgColor?: 'gray' | 'white';
}

/** Round button component for single-step user actions. */
export const RoundButton = ({
  label,
  hideLabel = false,
  onClick,
  disabled = false,
  selected = false,
  className,
  icon,
  size = 'md',
  bgColor = 'gray',
}: RoundButtonProps) => {
  const buttonSizeClass = tidyClasses(`
    ${size === 'sm' ? 'ds-size-[32px]' : ''}
    ${size === 'md' ? 'ds-size-[40px]' : ''}
    ${size === 'lg' ? 'ds-size-[64px]' : ''}
  `);

  const bgColorClass = tidyClasses(`
    ${!selected && bgColor === 'gray' ? 'ds-bg-bg-gray-2' : ''}
    ${!selected && bgColor === 'white' ? 'ds-bg-white' : ''}
  `);

  return (
    <button
      aria-label={label}
      disabled={disabled}
      type="button"
      onClick={onClick}
      className={tidyClasses(`
        ${className ?? ''}
        ${disabled ? 'ds-cursor-not-allowed ds-opacity-50' : ''}
        ds-group
        ds-flex
        ds-flex-col
        ds-justify-center 
        ds-items-center
        ds-gap-2
      `)}
    >
      <div
        aria-hidden
        className={tidyClasses(`
          ${buttonSizeClass}
          ${selected ? 'ds-bg-accent' : bgColorClass}
          ${selected ? 'ds-text-white' : 'ds-text-black group-hover:ds-text-accent'}
          ds-rounded-full
          ds-flex
          ds-items-center
          ds-justify-center
          ds-select-none
      `)}
      >
        {icon}
      </div>
      <LabelPart label={label} hideLabel={hideLabel} selected={selected} />
    </button>
  );
};

const LabelPart = ({ label, hideLabel, selected }: Pick<RoundButtonProps, 'label' | 'hideLabel' | 'selected'>) => (
  <span
    className={tidyClasses(`
      ${hideLabel ? 'ds-hidden' : ''}
      ${selected ? 'ds-text-accent' : 'ds-text-black'}
      ds-text-button-sm-mobile sm:ds-text-button-md
      group-hover:ds-text-accent
      group-hover:ds-underline
    `)}
  >
    {label}
  </span>
);
