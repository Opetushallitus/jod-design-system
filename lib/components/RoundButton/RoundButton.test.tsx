import { afterEach, describe, expect, it, test, vi } from 'vitest';
import { screen, render, cleanup } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { RoundButton } from './RoundButton';

afterEach(() => {
  cleanup();
});

describe('Snapshot testing', () => {
  test('Default', () => {
    const { container } = render(<RoundButton label="Default" onClick={vi.fn()} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('selected', () => {
    const { container } = render(<RoundButton label="Selected" selected={true} onClick={vi.fn()} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('Disabled, selected', () => {
    const { container } = render(<RoundButton label="Disabled, non-selected" selected={true} onClick={vi.fn()} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('Disabled, non-selected', () => {
    const { container } = render(<RoundButton label="Disabled, non-selected" onClick={vi.fn()} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('Disabled, selected', () => {
    const { container } = render(<RoundButton label="Disabled, non-selected" selected={true} onClick={vi.fn()} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

it('has the correct label', () => {
  render(<RoundButton label="Label" onClick={vi.fn()} />);
  expect(screen.getByRole('button', { name: 'Label' })).not.toBeNull();
});

it('calls the callback on click', async () => {
  const mockCallback = vi.fn();
  const user = userEvent.setup();

  render(<RoundButton label="Test callback" onClick={mockCallback} />);
  const button = screen.getByRole('button', { name: 'Test callback' });
  await user.click(button);
  expect(mockCallback).toHaveBeenCalledTimes(1);
});

it('does not call the callback while disabled', async () => {
  const mockCallback = vi.fn();
  const user = userEvent.setup();

  render(<RoundButton label="Disabled RoundButton" onClick={mockCallback} disabled />);
  const button = screen.getByRole('button', { name: 'Disabled RoundButton' });
  await user.click(button);
  expect(mockCallback).toHaveBeenCalledTimes(0);
});

describe('disabled', () => {
  it('should add correct CSS classes', () => {
    render(<RoundButton label="Disabled RoundButton" disabled={true} onClick={vi.fn()} />);
    const parentClassList = screen.getByText('Disabled RoundButton').parentElement?.classList;
    expect(parentClassList).toContain('cursor-not-allowed');
    expect(parentClassList).toContain('opacity-50');
  });

  it('has the disabled attribute', () => {
    const { getByRole } = render(<RoundButton label="Disabled RoundButton" disabled={true} onClick={vi.fn()} />);
    expect(getByRole('button')).toHaveProperty('disabled', true);
  });
});
