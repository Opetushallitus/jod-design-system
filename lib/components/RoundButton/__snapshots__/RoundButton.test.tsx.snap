// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot testing > Default 1`] = `
<div
  class="flex flex-col items-center justify-center px-1"
>
  <button
    aria-labelledby=":r0:"
    class="size-[72px] rounded-full border border-transparent hover:ring hover:ring-purple-500 focus:outline-none focus:ring focus:ring-purple-500 focus:hover:border-none bg-[#f5f5f5]"
    type="button"
  >
    <span
      class="text-center text-5xl leading-6 text-[#4d5358]"
    >
      ☰
    </span>
  </button>
  <span
    class="text-sm font-normal"
    id=":r0:"
  >
    Default
  </span>
</div>
`;

exports[`Snapshot testing > Disabled, non-selected 1`] = `
<div
  class="flex flex-col items-center justify-center px-1"
>
  <button
    aria-labelledby=":r3:"
    class="size-[72px] rounded-full border border-transparent hover:ring hover:ring-purple-500 focus:outline-none focus:ring focus:ring-purple-500 focus:hover:border-none bg-[#f5f5f5]"
    type="button"
  >
    <span
      class="text-center text-5xl leading-6 text-[#4d5358]"
    >
      ☰
    </span>
  </button>
  <span
    class="text-sm font-normal"
    id=":r3:"
  >
    Disabled, non-selected
  </span>
</div>
`;

exports[`Snapshot testing > Disabled, selected 1`] = `
<div
  class="flex flex-col items-center justify-center px-1"
>
  <button
    aria-labelledby=":r2:"
    class="size-[72px] rounded-full border border-transparent hover:ring hover:ring-purple-500 focus:outline-none focus:ring focus:ring-purple-500 focus:hover:border-none bg-[#697077]"
    type="button"
  >
    <span
      class="text-center text-5xl leading-6 text-[#ffffff]"
    >
      ☰
    </span>
  </button>
  <span
    class="text-sm font-normal"
    id=":r2:"
  >
    Disabled, non-selected
  </span>
</div>
`;

exports[`Snapshot testing > Disabled, selected 2`] = `
<div
  class="flex flex-col items-center justify-center px-1"
>
  <button
    aria-labelledby=":r4:"
    class="size-[72px] rounded-full border border-transparent hover:ring hover:ring-purple-500 focus:outline-none focus:ring focus:ring-purple-500 focus:hover:border-none bg-[#697077]"
    type="button"
  >
    <span
      class="text-center text-5xl leading-6 text-[#ffffff]"
    >
      ☰
    </span>
  </button>
  <span
    class="text-sm font-normal"
    id=":r4:"
  >
    Disabled, non-selected
  </span>
</div>
`;

exports[`Snapshot testing > selected 1`] = `
<div
  class="flex flex-col items-center justify-center px-1"
>
  <button
    aria-labelledby=":r1:"
    class="size-[72px] rounded-full border border-transparent hover:ring hover:ring-purple-500 focus:outline-none focus:ring focus:ring-purple-500 focus:hover:border-none bg-[#697077]"
    type="button"
  >
    <span
      class="text-center text-5xl leading-6 text-[#ffffff]"
    >
      ☰
    </span>
  </button>
  <span
    class="text-sm font-normal"
    id=":r1:"
  >
    Selected
  </span>
</div>
`;
